roomserver/internal/input/input_events.go

```diff
@@ -87,24 +101,69 @@ func (r *Inputer) processRoomEvent(
 				switch idFormat {
 				case gomatrixserverlib.EventIDFormatV1:
 					if bytes.Equal(event.EventReference().EventSHA256, evs[0].EventReference().EventSHA256) {
-						util.GetLogger(ctx).WithField("event_id", event.EventID()).Infof("Already processed event; ignoring")
+						logger.Debugf("Already processed event; ignoring")
 						return nil
 					}
 				default:
-					util.GetLogger(ctx).WithField("event_id", event.EventID()).Infof("Already processed event; ignoring")
+					logger.Debugf("Already processed event; ignoring")
 					return nil
 				}
 			}
 		}
 	}
 
-	// Check that the event passes authentication checks and work out
-	// the numeric IDs for the auth events.
+	missingRes := &api.QueryMissingAuthPrevEventsResponse{}
+	serverRes := &fedapi.QueryJoinedHostServerNamesInRoomResponse{}
+	if event.Type() != gomatrixserverlib.MRoomCreate {
+		missingReq := &api.QueryMissingAuthPrevEventsRequest{
+			RoomID:       event.RoomID(),
+			AuthEventIDs: event.AuthEventIDs(),
+			PrevEventIDs: event.PrevEventIDs(),
+		}
+		if err = r.Queryer.QueryMissingAuthPrevEvents(ctx, missingReq, missingRes); err != nil {
+			return fmt.Errorf("r.Queryer.QueryMissingAuthPrevEvents: %w", err)
+		}
+	}
+	if len(missingRes.MissingAuthEventIDs) > 0 || len(missingRes.MissingPrevEventIDs) > 0 {
+		serverReq := &fedapi.QueryJoinedHostServerNamesInRoomRequest{
+			RoomID: event.RoomID(),
+		}
+		if err = r.FSAPI.QueryJoinedHostServerNamesInRoom(ctx, serverReq, serverRes); err != nil {
+			return fmt.Errorf("r.FSAPI.QueryJoinedHostServerNamesInRoom: %w", err)
+		}
+	}
+	if input.Origin != "" {
+		serverRes.ServerNames = append([]gomatrixserverlib.ServerName{input.Origin}, serverRes.ServerNames...)
+	}
+	if len(serverRes.ServerNames) > 5 {
+		serverRes.ServerNames = serverRes.ServerNames[:5]
+	}
+
+	// First of all, check that the auth events of the event are known.
+	// If they aren't then we will ask the federation API for them.
 	isRejected := false
-	authEventNIDs, rejectionErr := helpers.CheckAuthEvents(ctx, r.DB, headered, input.AuthEventIDs)
-	if rejectionErr != nil {
-		logrus.WithError(rejectionErr).WithField("event_id", event.EventID()).WithField("auth_event_ids", input.AuthEventIDs).Error("helpers.CheckAuthEvents failed for event, rejecting event")
+	authEvents := gomatrixserverlib.NewAuthEvents(nil)
+	knownEvents := map[string]*types.Event{}
+	if err = r.checkForMissingAuthEvents(ctx, logger, headered, &authEvents, knownEvents, serverRes.ServerNames); err != nil {
```

Comment:
Given `checkForMissingAuthEvents` is called in exactly one place (here), and given we init empty `knownEvents` and `authEvents`, can we please just make this `fetchAuthEvents(ctx, logger, headered, serverNames) (authEvents, knownEvents, error)`. It's not even right to say _missing_ here as we _don't know_ if they are missing as we blindly call this function irrespective of `if len(missingRes.MissingAuthEventIDs) > 0 || len(missingRes.MissingPrevEventIDs) > 0 {` on `:127`.